typescript/cli/src/fork/fork.ts

```diff
@@ -0,0 +1,302 @@
+
+export const ForkedChainConfigByChainSchema = z.record(ForkedChainConfigSchema);
+
+export type ForkedChainConfigByChain = z.infer<
+  typeof ForkedChainConfigByChainSchema
+>;
+
+export async function runForkCommand({
+  context,
+  chainsToFork,
+  forkConfig,
+  kill,
+  basePort = 8545,
+}: {
+  context: CommandContext;
+  chainsToFork: Set<ChainName>;
+  forkConfig: ForkedChainConfigByChain;
+  kill: boolean;
+  basePort?: number;
+}): Promise<void> {
+  const filteredChainsToFork = Array.from(chainsToFork).filter(
+    (chain) =>
+      context.multiProvider.getProtocol(chain) === ProtocolType.Ethereum,
+  );
+
+  let port = basePort;
+  for (const chainName of filteredChainsToFork) {
+    await forkChain(
+      context.multiProvider,
+      chainName,
+      port,
+      kill,
+      forkConfig[chainName],
+    );
+
+    port++;
+  }
+}
+
+async function forkChain(
+  multiProvider: MultiProvider<{}>,
+  chainName: ChainName,
+  forkPort: number,
+  kill: boolean,
+  forkConfig?: ForkedChainConfig,
+): Promise<void> {
+  let killAnvilProcess: ((isPanicking: boolean) => Promise<void>) | undefined;
+  try {
+    const chainMetadata = await multiProvider.getChainMetadata(chainName);
+
+    const rpcUrl = chainMetadata.rpcUrls[0];
+    if (!rpcUrl) {
+      logRed(`No rpc found for chain ${chainName}`);
+      process.exit(0);
+    }
+
+    const endpoint = `${LOCAL_HOST}:${forkPort}`;
+    logGray(`Starting Anvil node for chain ${chainName} at port ${forkPort}`);
+    const anvilProcess = $`anvil --port ${forkPort} --chain-id ${chainMetadata.chainId} --fork-url ${rpcUrl.http}`;
+
+    const provider = new JsonRpcProvider(endpoint);
+    await retryAsync(() => provider.getNetwork(), 10, 500);
+
+    logGray(
+      `Successfully started Anvil node for chain ${chainName} at ${endpoint}`,
+    );
+
+    killAnvilProcess = (isPanicking: boolean) =>
+      anvilProcess.kill(isPanicking ? 'SIGTERM' : 'SIGINT');
+    process.once('exit', () => killAnvilProcess && killAnvilProcess(false));
+
+    if (!forkConfig) {
+      return;
+    }
+
+    await handleImpersonations(
+      provider,
+      chainName,
+      forkConfig.impersonateAccounts,
+    );
+
+    await handleTransactions(provider, chainName, forkConfig.transactions);
+
+    if (kill) {
+      await killAnvilProcess(false);
+    }
+  } catch (error) {
+    // Kill any running anvil process otherwise the process will keep running
+    // in the background.
+    if (killAnvilProcess) {
+      await killAnvilProcess(true);
+    }
+
+    throw error;
+  }
+}
+
+async function handleImpersonations(
+  provider: JsonRpcProvider,
+  chainName: ChainName,
+  accountsToImpersonate: Address[],
+): Promise<void> {
+  if (accountsToImpersonate.length === 0) {
+    return;
+  }
+
+  logGray(
+    `Impersonating accounts ${accountsToImpersonate} on chain ${chainName}`,
+  );
+  await Promise.all(
+    accountsToImpersonate.map((address) =>
```

Comment:
as discussed, would be an improvement to simplify the config by removing this. we would just preprocess the `transactions.from` into this array if batch impersonations are absolutely necessary.